File: Python/277__Find_the_Celebrity.py
# The knows API is already defined for you.
# @param a, person a
# @param b, person b
# @return a boolean, whether a knows b
# def knows(a, b):

class Solution(object):
    def findCelebrity(self, n):
        """
        :type n: int
        :rtype: int
        """
        candidate = 0
        # if candidate knows anyone then it should be replaced
        # i is known by someone, so it might be a candidate
        # the last candidate is the one who dont konw anyone after it
        for i in range(n):
            if knows(candidate, i):
                candidate = i

        # check if the candidate know anyone before
        for i in range(candidate):
            if knows(candidate,i):
                return -1

        # check anyone after know candidate
        for i in range(candidate, n):
            if not knows(i, candidate):
                return -1

        return candidate

                

File: Java/277__Find_the_Celebrity.java
/* The knows API is defined in the parent class Relation.
      boolean knows(int a, int b); */

public class Solution extends Relation {
    public int findCelebrity(int n) {
        int[] num_i_know = new int[n];
        int[] known_by = new int[n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (knows(i, j)) {
                    num_i_know[i]++;
                    known_by[j]++;
                }
            }
        }
        for (int i = 0; i < n; i++) {
            if (num_i_know[i] == 1) {
                if (known_by[i] == n) {
                    return i;
                }
            }
        }
        return -1;
    }
}

// Smarter solution

/* The knows API is defined in the parent class Relation.
      boolean knows(int a, int b); */

public class Solution extends Relation {
    public int findCelebrity(int n) {
        int candidate = 0;

        for (int i = 1; i < n; i++) {
            if (knows(candidate, i)) {
                candidate = i;
            }
        }

        for (int i = 0; i < n; i++) {
            if (i == candidate) continue;
            if (knows(candidate, i) || !knows(i, candidate)) return -1;
        }

        return candidate;
    }
}
